core/src/compile/link.rs

```diff
@@ -0,0 +1,316 @@
+        self.0.new_slice()
+    }
+
+    pub(super) fn wrap(inner: G) -> Self {
+        Self(RecorderImpl {
+            inner,
+            tape: Rc::new(RefCell::new(Tape::new()))
+        })
+    }
+}
+
+impl<G, Y, R> Recorder<G, Y, R>
+    where
+        G: Generator<Yield = Y, Return = R>,
+{
+    fn complete<RR: Rng>(&mut self, rng: &mut RR) {
+        self.0.complete(rng)
+    }
+}
+
+pub(super) struct TapeViewImpl<Y, R> {
+    slice: SliceRef<Y, R>,
+    generation: usize,
+    range: Range<usize>
+}
+
+impl<Y, R> TapeViewImpl<Y, R> {
+    fn is_obsolete(&self) -> bool {
+        self.slice.get_generation() != self.generation
+    }
+
+    fn reset(&mut self) -> bool {
+        self.generation = self.slice.get_generation();
+        self.range = self.slice.new_range();
+        !self.range.is_empty()
+    }
+}
+
+impl<Y, R> Generator for TapeViewImpl<Y, R>
+where
+    GeneratorState<Y, R>: Clone,
+{
+    type Yield = Y;
+    type Return = Option<R>;
+
+    fn next<RR: Rng>(&mut self, rng: &mut RR) -> GeneratorState<Self::Yield, Self::Return> {
+        if self.is_obsolete() && !self.reset() {
+            return GeneratorState::Complete(None);
+        }
+
+        if let Some(idx) = self.range.next() {
+            (*self.slice.tape)
+                .borrow()
+                .get(idx)
+                .expect("a view's index was out of bound")
+                .clone()
+                .map_complete(Some)
+        } else if !self.reset() {
+            GeneratorState::Complete(None)
+        } else {
+            self.next(rng)
+        }
+    }
+}
+
+pub struct TapeView<Y, R>(pub(super) TapeViewImpl<Y, R>);
+
+pub(super) struct OrderedImpl<G, Y, R> {
+    is_complete: bool,
+    scope: Vec<SliceRef<Y, R>>,
+    children: Vec<(String, Recorder<G, Y, R>)>,
+    src: G,
+}
+
+impl<G, Y, R> Generator for OrderedImpl<G, Y, R>
+    where
+        G: Generator<Yield = Y, Return = R>,
+{
+    type Yield = Y;
+    type Return = R;
+
+    fn next<RR: Rng>(&mut self, rng: &mut RR) -> GeneratorState<Self::Yield, Self::Return> {
+        if !self.is_complete {
+            self.scope.iter_mut().for_each(|slice| {
+                slice.reset()
+            });
+            self.children.iter_mut().for_each(|(_, recorder)| {
+                recorder.complete(rng);
+            });
+            self.is_complete = true;
+        }
+        let state = self.src.next(rng);
+        self.is_complete = !state.is_complete();
+        state
+    }
+}
+
+pub struct Ordered<G, Y, R>(pub(super) OrderedImpl<G, Y, R>);
+
+impl<G, Y, R> Ordered<G, Y, R> {
+    pub(super) fn new<S, C>(scope: S, children: C, src: G) -> Self
+    where
+        S: IntoIterator<Item = SliceRef<Y, R>>,
+        C: IntoIterator<Item = (String, Recorder<G, Y, R>)>
+    {
+        Self(OrderedImpl {
+            is_complete: false,
+            scope: scope.into_iter().collect(),
+            children: children.into_iter().collect(),
+            src
+        })
+    }
+}
+
```

Comment:
Perhaps add a little documentation explaining taht each variant has to do with being a scope (ordered), a ref from(recorder) and ref to (view).